001-finding_resources.md

```diff
@@ -0,0 +1,120 @@
+# Design Document 001 - Finding Resources
+
+TODO: Overview
+
+## Finding SDF URI Resources
+
+The following sections discuss how `<uri>` elements in SDF files are
+resovled in order to find meshes and models. This section does not cover
+finding other resources, such as plugins.
+
+### Current Process
+
+This section describes how Ignition Gazebo currently finds Mesh URI
+resources. SDF contains multiple instances of URI elements, but the current
+implementation of Gazebo handles only mesh URIs. The information contained
+here can be extrapolated to handle other types of resources, such as
+textures and audio files.
+
+A user of Ignition may not need or use Gazebo. If this describes you, then
+skip directly to the last step which describes how
+`ignition::common::findFile` can be used to resolve a resource.
+
+1. Ignition Gazebo registers an SDF find callback in the Server constructor.
+   This callback is triggered by SDF whenever a URI resource from an
+   `<include>` element cannot be found. The Ignition Gazebo find callback in
+   turn calls Fuel Tools to download the resource.
+
+2. After registering the callback, Ignition Gazebo then finds the world file
+   specified on the command line in a path contained in the
+   `IGN_GAZEBO_RESOURCE_PATH` environment variable using
+   `common::SystemPaths::FindFile`.
+
+3. If the world file is found, then SDF is used to load the world file. SDF
+   will only attempt to resolve URI resources contained withing the world
+   file that are in `<include>` elements.
+
+4. A SimulationRunner is then instantiated for each `sdf::World`. The
+   LevelManager, contained in each SimulationRunner, loads each `sdf::Model`
+   and creates the appropriate set of components. Plugin loading is conducted
+   during this step, but no mesh URI resource loading is conducted.
+
+5. Systems and other aspects of simulation, such as rendering, must then each
+   resolve URI resources contained within the loaded world and models.
+   The `Util::asFullPath` is a helper function that is used to convert a URI
+   into a system path. This helper function only alters URIs that have
+   a relative path. URIs that have a scheme or are a full path are
```

Comment:
URIs that have a relative path and also have a file path.

For example, for models and actors included through `<include>`, the relative path is resolved with respect to the `.sdf` file. But for `<model>` tags directly inside the world, relative paths can't be resolved.

Reply:
These lines are referring only to the `Util::asFullPath` function, which only modifies a relative path. I think this correct based on the function's documentation.